File: tools/wave/src/network/api/api-handler.js
class ApiHandler {
  sendJson(object, response, status = 200) {
    const data = JSON.stringify(object, null, 2);
    response.set("Content-Type", "Application/JSON");
    response.status(status).send(data);
  }

  parseQueryParameters(request) {
    let {
      token,
      path,
      types,
      timeout,
      hostname,
      tokens,
      api,
      reftoken,
      reftokens,
      resume,
      redirect,
      data,
      count,
      status
    } = request.query;
    token = token || null;
    path = path || null;
    types = types ? types.split(",") : null;
    timeout = timeout || null;
    hostname = hostname || null;
    api = api || null;
    tokens = tokens ? tokens.split(",") : null;
    reftoken = reftoken || "";
    reftokens = reftokens || "";
    resume = resume ? (resume !== "false" ? true : false) : false;
    data = data || "";
    redirect = redirect || redirect === "";
    count = count || null;
    status = status ? status.split(",").map(status => status.trim()) : null;
    return {
      token,
      path,
      types,
      testTimeout: timeout,
      hostname,
      api,
      tokens,
      reftoken,
      reftokens,
      refTokens: reftokens,
      resume,
      redirect,
      data,
      count,
      status
    };
  }

  parseUrl(request) {
    let url = request.url;
    if (url.indexOf("?") !== -1) {
      url = url.split("?")[0];
    }
    // remove /api prefix
    url = url.split("/");
    url = url.filter(part => part !== "");
    url.shift();
    return url;
  }

  sendFile({ response, blob, filePath, fileName }) {
    response.set(
      "Content-Disposition",
      'attachment;filename="' + fileName + '"'
    );
    if (blob) {
      response.send(blob);
      return;
    }
    response.sendFile(filePath);
  }

  sendZip(options) {
    options.response.set("Content-Type", "application/x-compressed");
    this.sendFile(options);
  }
}

module.exports = ApiHandler;

File: tools/wptserve/wptserve/wave/network/api/api_handler.py
from __future__ import absolute_import
import json


class ApiHandler(object):
    def send_json(self, data, response, status=200):
        json_string = json.dumps(data)
        response.content = json_string
        response.headers = [(u"Content-Type", u"application/json")]
        response.status = status

    def parse_uri(self, request):
        uri_parts = []
        for part in request.request_path.split(u"/"):
            if part == u"":
                continue
            uri_parts.append(part)
        return uri_parts

    def parse_query_parameters(self, request):
        if u"?" not in request.request_path: return {}
        query = request.request_path.split(u"?")[1]
        if query == u"": return {}
        key_value_pairs = []
        if u"&" in query:
            key_value_pairs = query.split(u"&")
        else:
            key_value_pairs.append(query)

        parsed_parameters = {}
        for key_value_pair in key_value_pairs:
            if u"=" not in key_value_pair:
                parsed_parameters[key_value_pair] = True
                continue
            key, value = key_value_pair.split(u"=")
            parsed_parameters[key] = value

        return parsed_parameters


